### ISSUE TITLE
Restore Purchase - iOS

### ISSUE THREAD

**Issue by anoopda** [dooboolab/react-native-iap#3]
Any updates on iOS restore purchase feature?

**Comment by JJMoon** [comment]
We wil update soon. Maybe before 2018.

**Comment by KjellConnelly** [comment]
@JJMoon @dooboolab 
Btw how do we restore purchases on Android? I know Google Play allows you to query for non consumable purchases, but then that leaves me with a few questions:

1) Is the idea to query for IAPs when the app loads, and restore functionality on each app load (saving locally of course so the app can work offline as well)?
2) Which function should I use? I see:
``
refreshPurchaseItemsAndroid()
getPurchasedItemsAndroid()
``
Maybe an example is in order for the readme?
3) Do these only work for Google Play? How about Amazon/others?

**Comment by dooboolab** [comment]
@KjellConnelly 
Answering you questions~
1. You need to query for IAPs at first. When you want to query items offline and you are sure that lists of items won't change~ you can save them in LocalStorage.
2. Sorry for that. As we told in head of this repo, we are interested in making IAP the same behavior between IOS and ANDOIRD platform. Therefore we missed out that tutorial officially. getPurchasedItemsAndroid() is a method to get a list of items you purchased in your app. refreshPurchaseItemsAndroid() is a method to refresh that item so you can buy the same item again.
3. We've only tested on Google Play. We've not tested on Amazon/others yet.

Hope my answer helps you.

Thank you.

**Comment by KjellConnelly** [comment]
@dooboolab Thanks for the reply :) Could you help me understand a bit better how the process works for Android? I am coming from an iOS background, so I will use iOS terms to try to make sense of this. 

Imagine the situation where I have two IAPs
1) Premium Version ($3.99), non-consumable.
2) 100 Coins ($0.99), consumable.

In iOS, I would first query for the product IDs, then call the buy function. This returns Promise once the buy function has completed (success or failure).

On iOS, I would do something with LocalStorage such as setting a saved object:
```javascript
if (item.productId == 'com.website.100coins') {
    savedData = {
        coins: savedData.coins + 100,
        premium: savedData.premium
    }
} else if (item.productId == 'com.website.premium') {
    savedData = {
        coins: savedData.coins,
        premium: true
    }
}
```

And then I would have a Restore button, and when you tap it, the Restore function is called, which returns the non-consumable purchases that the user has made in the past using that iTunes account.

----------------------------

For Android, it looks like I query for the productIds just as on iOS. And when the purchase happens, I also save to LocalStorage. But if I remember correctly, I need to then tell Google Play that the consumable has been consumed (for the coins, and once I save to local storage). 

But when it comes to the Restore Button, I'm a bit confused.

By your description, it looks like those two functions do the following:

**getPurchasedItemsAndroid()**: This will return probably an array of IAPs that this user has purchased using your app. This will include all consumable, and non-consumable purchases. In my case, where I have Coins and a Premium Version, I would just iterate through this array to check to see if the user has bought the Premium Version before when they tap the Restore Button, right? And if its there, then I just set ```savedData.premium = true```?

**refreshPurchasedItemsAndroid()** I don't really understand what refreshing does in this case. Does it tell Google Play that all of the previously consumed purchases are still purchased, but not consumed anymore? Or does it refresh the app's list of purchased items from Google Play's servers in which case would mean that the user's device is saving some sort of data on its own when a purchase is made?

Finally, it seems like on iOS, you are required to have a Restore Button for non-consumables. But for Google Play, you CAN have a Restore button for non-consumables, but it would be best practice to just check using ```getPurchasedItemsAndroid``` every time the app is launched (if the user's local data is saying that they haven't purchased the premium version yet).

Sorry for the long essay. Just having trouble making sense of the IAP process on Android.

**Comment by dooboolab** [comment]
@KjellConnelly 

Sorry that I've missed this. I've not implemented Sku.Type option to get consumable products in android since implementation from ios was delayed. Before you could only getPurchasedItems lists for non-consumable products only. I just fixed this to support for querying both consumable and non-consumable products.
--------------------------- screenshots below -----------------------------------
1. In index.js
![screen1](https://user-images.githubusercontent.com/27461460/34101894-9a9c9202-e42a-11e7-967b-cc7c39a5fed4.png)
2. in RNIapModule.java
![screen2](https://user-images.githubusercontent.com/27461460/34101898-9e1247ba-e42a-11e7-84b7-b0cbca554b07.png)
--------------------------- screenshots done -----------------------------------


Also what refreshPurchasedItemsAndroid I've updated support for conusmable products. Screenshots are below
--------------------------- screenshots below -----------------------------------
1. In index.js
![screen1](https://user-images.githubusercontent.com/27461460/34102281-dd51b69e-e42b-11e7-8cca-abacf8be7c1e.png)
2. in RNIapModule.java
![screen2](https://user-images.githubusercontent.com/27461460/34102285-e22bd370-e42b-11e7-984b-759ab57d8dba.png)
--------------------------- screenshots done -----------------------------------

I've just commited these updates. Hope this helps your question.

Now you can call getPurchasedItemsAndroid(type: string) function to get non-consumable items or consumable items. If you want non-consumable items just put null or any string. If you want consumable items put 'SUBS' in it.

You can use refreshPurchasedItemsAndroid(type: string) function same as described above.

**Comment by dooboolab** [comment]
@KjellConnelly 

Yes! It was about items for subscriptions. OMG... I was confused too. Thanks again to point that out. Sorry that I was confused by busy work ..

It seems we've been missing some support for this senario. We didn't actually distinguish between consumable and non-consumable product in our library. Since you need this part, we will support for this near future. Thanks for sharing.

**Comment by KjellConnelly** [comment]
@dooboolab If you have time, could you add Restore Purchase support for iOS? Will be needing it soon and was hoping this feature to be implemented by 2018. Thanks!

**Comment by JJMoon** [comment]
@KjellConnelly 
Sorry for delay. I will implement that method soon. Happy new year.. !!

**Comment by KjellConnelly** [comment]
Thank you!! I really appreciate it :)

**Comment by dooboolab** [comment]
@KjellConnelly 
Checkout later version of @0.2.2 for update.
Thank you!

**Comment by ivan133** [comment]
But how to use it?
